### PR TITLE
#1185 Macromolecules export from MolV3000 to KET

### PR DESCRIPTION
**Pull request by even1024**

### CODE REVIEW

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 1]
```diff
@@ -1066,26 +1066,25 @@ std::string MoleculeJsonSaver::monomerId(const TGroup& tg)
...
     return name;
 }
 
-std::string MoleculeJsonSaver::monomerHELMClass(const TGroup& tg)
+std::string MoleculeJsonSaver::monomerHELMClass(const std::string& class_name)
 {
-    auto mclass = std::string(tg.tgroup_class.ptr());
-    if (kAminoClasses.find(mclass) != kAminoClasses.end())
+    if (kAminoClasses.find(class_name) != kAminoClasses.end())
         return "PEPTIDE";
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1078]: make constant for PEPTIDE

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 2]
```diff
@@ -1066,26 +1066,25 @@ std::string MoleculeJsonSaver::monomerId(const TGroup& tg)
...
-    if (kNucleicClasses.find(mclass) != kNucleicClasses.end())
+    if (kNucleicClasses.find(class_name) != kNucleicClasses.end())
         return "RNA";
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1080]: make constant for RNA

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 3]
```diff
@@ -1066,26 +1066,25 @@ std::string MoleculeJsonSaver::monomerId(const TGroup& tg)
...
     return "CHEM";
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1081]: make constant for CHEM

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 4]
```diff
@@ -1066,26 +1066,25 @@ std::string MoleculeJsonSaver::monomerId(const TGroup& tg)
...
 }
 
-std::string MoleculeJsonSaver::monomerKETClass(const TGroup& tg)
+std::string MoleculeJsonSaver::monomerKETClass(const std::string& class_name)
 {
-    auto mclass = std::string(tg.tgroup_class.ptr());
-    if (mclass == "AA")
+    auto mclass = class_name;
+    if (class_name == "AA")
         return "AminoAcid";
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1088]: make constant for AminoAcid and AA

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 5]
```diff
@@ -1066,26 +1066,25 @@ std::string MoleculeJsonSaver::monomerId(const TGroup& tg)
...
 
     if (mclass == "dAA")
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1090]: make constant for dAA

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 6]
```diff
@@ -1100,12 +1099,12 @@ std::string MoleculeJsonSaver::monomerKETClass(const TGroup& tg)
...
 
-std::string MoleculeJsonSaver::naturalAnalog(const TGroup& tg)
+std::string MoleculeJsonSaver::naturalAnalog(const std::string& natreplace)
 {
     std::string res;
-    if (tg.tgroup_natreplace.ptr())
+    if (natreplace.size())
     {
-        auto nat_replace = split(std::string(tg.tgroup_natreplace.ptr()), '/');
+        auto nat_replace = split(std::string(natreplace.c_str()), '/');
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1107]: this logic should not be in json saver

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 7]
```diff
@@ -1139,12 +1138,81 @@ std::string MoleculeJsonSaver::monomerAlias(const TGroup& tg)
     return alias;
 }
 
+void MoleculeJsonSaver::saveSuperAtomAsTemplate(BaseMolecule& mol, int sa_index, int template_number, JsonWriter& writer)
+{
+    Superatom& sa = (Superatom&)mol.sgroups.getSGroup(sa_index);
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1143]: this logic should not be part of JSON saver. Super atoms should be converted to templates while reading from molfile

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 8]
```diff
@@ -1139,12 +1138,81 @@ std::string MoleculeJsonSaver::monomerAlias(const TGroup& tg)
...
+        auto& tg = temp_it->second.get();
+        if (template_class.empty() && tg.tgroup_class.size())
+            template_class = tg.tgroup_class.ptr();
+        if (alias.empty())
+            alias = monomerAlias(tg);
+    }
+
+    // if failed to determine the class, fallback to CHEM
+    if (template_class.empty())
+        template_class = "CHEM";
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1180]: add constant for CHEM

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 9]
```diff
@@ -1187,72 +1255,66 @@ void MoleculeJsonSaver::saveMonomerTemplate(TGroup& tg, JsonWriter& writer)
     writer.EndObject();
 }
 
+void MoleculeJsonSaver::saveSuperatomAttachmentPoints(Superatom& sa, JsonWriter& writer)
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1258]: this logic should not be part of json saver

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 10]
```diff
@@ -1316,48 +1378,139 @@ bool MoleculeJsonSaver::_checkAttPointOrder(BaseMolecule& mol, int rsite)
...
+        auto& tg = mol.tgroups.getTGroup(i);
+        _templates.emplace(tg.tgroup_name.size() ? tg.tgroup_name.ptr() : monomerAlias(tg), std::ref(tg));
+    }
+}
 
-    std::unique_ptr<BaseMolecule> mol(bmol.neu());
-    mol->clone_KeepIndices(bmol);
+void MoleculeJsonSaver::collectSCSRSuperAtoms(BaseMolecule& mol)
+{
+    _scsr_atom_superatoms.clear();
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1394]: remove _scsr_atom_superatoms - it should be converted to templates while loading from molfile

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 11]
```diff
@@ -1370,34 +1523,225 @@ void MoleculeJsonSaver::saveMolecule(BaseMolecule& bmol, JsonWriter& writer)
...
     {
+        // collect attachment points into unordered map <key, val>. keu - pair of from and destination atom. val - attachment point name.
+        _monomer_connections.clear();
+        for (int i = mol.template_attachment_points.begin(); i != mol.template_attachment_points.end(); i = mol.template_attachment_points.next(i))
+        {
+            auto& sap = mol.template_attachment_points.at(i);
+            _monomer_connections.emplace(std::make_pair(sap.ap_occur_idx, sap.ap_aidx), sap.ap_id.ptr());
+        }
+
+        for (const auto& kvp : _scsr_superatoms)
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1561]: there should be no _scsr_superatoms
all the modified monomers should be loaded into TGroups while loading from molfile

File: core/indigo-core/molecule/src/molecule_json_saver.cpp [hunk 12]
```diff
@@ -1370,34 +1523,225 @@ void MoleculeJsonSaver::saveMolecule(BaseMolecule& bmol, JsonWriter& writer)
...
+                if (tg_it != _templates.end())
+                {
+                    writer.Key("templateId");
+                    writer.String(monomerId(tg_it->second.get()).c_str());
+                }
+                writer.EndObject(); // monomer
+            }
+        }
+
+    for (const auto& kvp : _scsr_superatoms)
```
Comment (AlexanderSavelyev) [inline core/indigo-core/molecule/src/molecule_json_saver.cpp:1691]: remove _scsr_superatoms logic

**Review by AlexanderSavelyev — CHANGES_REQUESTED** [review]

File: core/indigo-core/molecule/src/molecule_json_saver.cpp (hunk 2 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_json_saver.cpp]: Fixed

File: core/indigo-core/molecule/src/molecule_json_saver.cpp (hunk 4 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_json_saver.cpp]: Fixed

Comment (even1024) [reply core/indigo-core/molecule/src/molecule_json_saver.cpp]: Fixed

File: core/indigo-core/molecule/src/molecule_json_saver.cpp (hunk 5 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_json_saver.cpp]: Fixed

File: core/indigo-core/molecule/src/molecule_json_saver.cpp (hunk 9 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_json_saver.cpp]: Fixed

File: core/indigo-core/molecule/src/molecule_json_saver.cpp (hunk 10 above)
Comment (even1024) [reply core/indigo-core/molecule/src/molecule_json_saver.cpp]: Fixed

**Review by AlexanderSavelyev — APPROVED** [review]

**Pull request merged**